---
title: ''
output: html_document
params:
  ovr: 0
  cpu: !r parallel::detectCores()
---

```{r setup, include=FALSE}
library(readr)
library(dplyr)
library(tidyr)
library(stringr)
library(tibble)
library(IRanges)
library(lubridate)
library(BactDating)
library(ggplot2)
library(ggtree)
do.call(Sys.setenv, params)
```

```{r}
parse_8C <- function(path) {
  i <- 0
  col_types <- cols(
    `% identity` = "d", `alignment length` = "i",
    mismatches = "i", `gap opens` = "i",
    `q. start` = "i", `q. end` = "i",
    `s. start` = "i", `s. end` = "i",
    evalue = "d", `bit score` = "d",
    .default = "c"
  )
  header <- read_lines(path, n_max = 6)
  qlen <- as.integer(tail(strsplit(header[3], " ")[[1]], 2)[1])
  fields <- strsplit(substr(header[5], 11, nchar(header[5])), ", ")[[1]]
  hits <- cbind(
    read_tsv(
      path,
      col_names = fields,
      col_types = col_types,
      comment = "#"
    ),
    `query length` = qlen
  )
  hits
}
```

```{r}
df.qaccver <-
  read_tsv("query.tsv", comment = "#", col_types = "cdddccc") %>%
  mutate(accver = ifelse(
    strand == 1,
    paste0(accver, ":", start, "-", stop),
    paste0(accver, ":c", stop, "-", start)
  ))
df.genome <- read_tsv(
  "genome.tsv",
  col_types = "ccifffffcccccccddcc", na = "?"
)
```

```{r}
Sys.glob("[A-G]-*/hits.tsv") %>%
  lapply(parse_8C) %>%
  bind_rows() %>%
  mutate(
    qpidcov = abs(`q. end` - `q. start`) / `query length` * `% identity`,
    strand = if_else(`q. start` < `q. end`, "plus", "minus")
  ) %>%
  group_by(`query id`, `subject id`) %>%
  slice_max(qpidcov) %>%
  ungroup() %>%
  left_join(
    select(df.qaccver, accver, species, region),
    by = c(`query id` = "accver")
  ) ->
df.qpidcov
```

```{r}
select(df.qpidcov, `subject id`, qpidcov, species, region) %>%
arrange(species, region) %>%
pivot_wider(
  names_from = c("species", "region"),
  names_sep = "-",
  values_from = "qpidcov"
)
```

```{r}
left_join(
  df.qpidcov,
  select(df.genome, accver, species, date, ac0, latitude, longitude),
  by = c(`subject id` = "accver")
) %>%
  filter(!(region %in% c("fib1", "fib2"))) %>%
  filter(species.y != "G") %>%
  filter(!is.na(species.y)) %>%
  filter(!is.na(date)) %>%
  filter(!is.na(ac0)) %>%
  dplyr::rename(country = ac0) %>%
  group_by(`subject id`, region) %>%
  slice_max(qpidcov, n = 1) %>%
  ungroup() %>%
  group_by(species.x, region, strand) %>%
  group_walk(function(.x, .y) {
    root <- paste0(.y[1:2], collapse = "-")
    write_tsv(
      mutate(.x, strain = `subject id`),
      file.path(root, "meta.tsv")
    )
    write_tsv(
      cbind("country", distinct(.x[c("country", "latitude", "longitude")])),
      file.path(root, "coor.tsv"),
      col_names = F
    )
    write_lines(
      with(.x, sprintf("%s:%d-%d", `subject id`, `s. start`, `s. end`)),
      file.path(root, paste0("reg", "-", .y[3], ".txt"))
    )
  }) ->
df.reg
```

```{r, fig.height=20, fig.width=10}
df.reg <- df.reg[order(ordered(df.reg$species.x, levels = c("A", "B", "C", "D", "E", "F", "G"))), ]
df.reg$`subject id` <- factor(df.reg$`subject id`, levels = unique(df.reg$`subject id`))
ggplot(df.reg) +
  geom_segment(aes(x = `s. start`, xend = `s. end`, y = `subject id`, yend = `subject id`, color = species.x)) +
  facet_wrap( ~ region, scales = "free_x", nrow = 1) +
  theme_minimal()

ggsave("blast.pdf")
```

```{bash}
find . -name "[A-F]-*" -type d | \
while read -r ele
do
  {
    [[ -f "$ele/reg-plus.txt" ]] && samtools faidx -r "$ele/reg-plus.txt" blast/genome.fasta
    [[ -f "$ele/reg-minus.txt" ]] && samtools faidx -r "$ele/reg-minus.txt" -i --mark-strand no blast/genome.fasta
  } > "$ele/seq.fasta"
  mafft --auto --thread "$cpu" "$ele/seq.fasta" > "$ele/msa.fasta" 2> "$ele/msa.log"
  sed -i '' -e '/^>/ s/:.*//g' -e '/^[^>]/ s/[^ACGTNacgtn-]/n/g' "$ele/msa.fasta"
  run_gubbins.py "$ele/msa.fasta" -p "$ele/gub" -c "$cpu" > "$ele/gub.log"
done
```

```{r, fig.height=4, fig.width=10}
tree <- loadGubbins("B-gen/bat.gub")
tip.date <- decimal_date(ymd(with(df.genome, date[match(tree$tip.label, accver)])))
tip.type <- as.integer(with(df.genome, type[match(tree$tip.label, accver)]))
tip.geo <- with(df.genome, ac0[match(tree$tip.label, accver)])
tree <- initRoot(tree, tip.date, useRec = T)
roottotip(tree, tip.date)
```

```{r}
res <- bactdate(tree, tip.date, nbIts = 250000, thin = 10, model = "relaxedgamma", useRec = T, showProgress = T)
```

```{r}
plot(res, "trace")
```

```{r}
coda::effectiveSize(as.mcmc.resBactDating(res))
```

```{r, fig.width=10}
chron <- as.treedata.resBactDating(res)
chron <- methods::new(
  "treedata",
  phylo = chron[[1]],
  data = tibble::as_tibble(chron[[2]])
)
chron@phylo <- groupOTU(chron@phylo, split(chron@phylo$tip.label, tip.type), "type")
ggtree(chron, mrsd = decimal2Date(min(tip.date)), size = 0.75) +
  aes(color = type) +
  geom_range("height_0.95_HPD", alpha = 0.5, size = 1) +
  theme_tree2() +
  theme(
    legend.position = "bottom",
    panel.grid.major.x = element_line(color = "black", size = .25),
    panel.grid.minor.x = element_line(color = "grey", size = .25)
  )
```
